Feature/m0081 meja 1638 document portable fo (#375)
MicroEJ/docs: ApplicationDeveloperGuide/soar.rst

to ``T1``.

.. _soar_method_devirtualization:

Method Devirtualization
------------------------

Method devirtualization consists of transforming a virtual method call to a direct method call when possible.

> A virtual method call is a call to a method declared either in an interface or in a class when it not declared with private visibility. 

Author: the `final` flag has no impact on the virtualization optimizations of the JDT (the flag can be modified later)
![image](https://user-images.githubusercontent.com/19801840/157429806-9c45d594-bd14-4ab7-8edf-b5f56184552c.png)

> A virtual method call is a call to a method declared either in an interface or in a class when it not declared with private visibility. 

Reviewer: ok. maybe this info can be added to the doc

Calls to constructors and private methods are already as a direct method call by the Java compiler.

> SOAR automatically transforms virtual method call to direct method call if there is one and only one embedded implementation method.

Author: it throws AbstractMethodError at runtime accorfing to Java specification

> SOAR automatically transforms virtual method call to direct method call if there is one and only one embedded implementation method.

Reviewer: ok

.. note::
  
   The list of embedded methods is available in the :ref:`SOAR Information File <soar_info_file>`.

.. _soar_method_inlining:

Method Inlining
---------------

Method inlining consists of replacing a direct method call by the content of the method. This avoids the creation of a new stack frame context which can be slower than executing the code itself.
Method inlining is transitively applied from leaf to root methods.

The following code patterns are inlined:

> - empty code after processing :ref:`assertions <enable_assertions_emb>` and :ref:`if code removal <if_constant_removal>`.

Author: Added "The following **method** code patterns. "

- call to a constructor with no parameters.

> - call to a private method with no parameters.

Author: I added a note

> - call to a private method with no parameters.

Author: after double check, the inline pattern detection is done in an early phase by SOAR-C before devirtualization with is done in an late phase by SOAR-O.
So the JDT must have generated invoke special (constructor or private method)

> - call to a private method with no parameters.

Reviewer: ok so the 2 optimizations are not chained? why not run devirtualization before inlining?

> - call to a private method with no parameters.

Author: the inline pattern detection is done in an early phase by SOAR-C, while devirtualization is done in a late phase by SOAR-O

> - call to a private method with no parameters.

Reviewer: yes my question is could the inline detection be performed later? or devirtualization be done earlier?

> - call to a private method with no parameters.

Author: I don't think this can be changed easily.

> - call to a private method with no parameters.

Reviewer: ok

to ``T1``.

.. _soar_method_devirtualization:

Method Devirtualization
------------------------

Method devirtualization consists of transforming a virtual method call to a direct method call when possible.
A virtual method call is a call to a non private method declared either in an interface or in a class. 
A call to a constructor or a private method is already optimized as a direct method call by the Java compiler.

SOAR automatically optimizes a virtual method call to direct method call if there is one and only one embedded implementation method.

.. note::
  

>    The list of embedded methods is available in the :ref:`SOAR Information File <soar_info_file>`.

Reviewer 2: ```suggestion
   SOAR generates the list of the embedded methods in the :ref:`SOAR Information File <soar_info_file>`.
```

.. _soar_method_inlining:

Method Inlining
---------------

> Method inlining consists of replacing a direct method call by the content of the method. This avoids the creation of a new stack frame context which can be slower than executing the code itself.

Reviewer 2: ```suggestion
Method inlining consists of replacing a direct method call with the content of the method. This avoids the creation of a new stack frame context, which can be slower than executing the code itself.
```

Method inlining is transitively applied from leaf to root methods.

The following method code patterns are inlined:

- empty code after processing :ref:`assertions <enable_assertions_emb>` and :ref:`if code removal <if_constant_removal>`.
- call to a constructor with no parameters.
- call to a private method with no parameters.
- call to a static method with no parameters, if and only if the caller is also a static method.

.. note::

   Method inlining is performed after :ref:`method devirtualization <soar_method_devirtualization>`, so a virtual method call will be inlined 

>    if there is a unique embedded implementation method that matches one of inlined method code patterns.

Reviewer 2: ```suggestion
   if there is a unique embedded implementation method that matches one of the inlined method code patterns.
```
